### PR TITLE
Remove warnings, outdated templates, proxy commands to C3

### PR DESCRIPTION
**Pull request by irvinebroque**
## What this PR solves / how to test

1. Removes the outdated directory of templates in [`/templates`](https://github.com/cloudflare/workers-sdk/tree/main/templates). None of these are actually up-to-date. As-needed some of this content can be added to `npm create cloudflare` templates that live [here](https://github.com/cloudflare/workers-sdk/tree/main/packages/create-cloudflare/templates).
2. `wrangler generate` and `wrangler init` both just kick you over to `npm create cloudflare` automatically now. `wrangler generate template <repository_url>` behavior still works, just now via create cloudflare.
3. Removes a lot of old code

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required / Maybe required
  - [ ] Not required because: e2e doesn't cover pipelines
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because: 
- Public documentation
  - [ ] TODO (before merge)
  - [x] Cloudflare docs PR(s): Draft: https://github.com/cloudflare/cloudflare-docs/pull/16825
  - [ ] Not necessary because: pre release beta feature## What

### CODE REVIEW

File: packages/wrangler/src/generate/index.ts
Comment (irvinebroque) [inline packages/wrangler/src/generate/index.ts:43]: Need to rip out `initHandler` too and just make it simply proxy over to C3. Bunch of things to rip out there too.

**Review by Cherry — COMMENTED** [review]
Excited to see this cleanup! It's always been confusing to explain to people the difference between `init`, `generate`, and the new c3. 

However, since wrangler follows semantic versioning, and `generate` was never beta/experimental, this is going to to land in a v4, right so to not break existing workflows? `wrangler generate test /some/company/git/url` is still very useful today.

The docs at https://developers.cloudflare.com/workers/wrangler/deprecations/ even mention `generate` specifically and affirm it won't be removed until v4.

File: packages/wrangler/src/generate/index.ts
Comment (Skye-31) [inline packages/wrangler/src/generate/index.ts]: Why not call C3 with the template that the user specifies and let it do it's flow?

**Comment by irvinebroque** [comment]
Line numbers on errors that `pnpm run build` outputs are confusing to me, and this is weird, this part not modified.

```
wrangler:build: src/index.ts(339,3): error TS2769: No overload matches this call.
wrangler:build:   The last overload gave the following error.
wrangler:build:     Argument of type '(yargs: CommonYargsArgv) => Argv<CommonYargsOptions & { name: string | undefined; } & { type: string | undefined; } & { site: boolean | undefined; } & { ...; } & { ...; } & { ...; }>' is not assignable to parameter of type '{ [key: string]: Options; }'.
wrangler:build:       Index signature for type 'string' is missing in type '(yargs: CommonYargsArgv) => Argv<CommonYargsOptions & { name: string | undefined; } & { type: string | undefined; } & { site: boolean | undefined; } & { ...; } & { ...; } & { ...; }>'.
wrangler:build: src/index.ts(782,3): error TS2769: No overload matches this call.
wrangler:build:   The last overload gave the following error.
wrangler:build:     Argument of type 'boolean' is not assignable to parameter of type 'string'.
```

File: .changeset/gold-toys-smell.md
Comment (Cherry) [inline .changeset/gold-toys-smell.md:2]: If this is to land in a minor and just passes the args to c3, behaviour between c3's git init and generate needs to match 1:1. I've opened https://github.com/cloudflare/workers-sdk/pull/6415 for a bunch of differences and issues I've found with c3's implementation.

File: packages/wrangler/src/generate/index.ts
Comment (andyjessop) [inline packages/wrangler/src/generate/index.ts:107]: @irvinebroque as discussed, this change will delegate to c3 if a Cloudflare folder template is provided.

What I haven't done is pass through the template name itself, because there doesn't seem to be much overlap between the two, and if you provide a template that doesn't exist, it will error. We'll therefore have a confusing situation if it will occasionally succeed and occasionally not. For the sake of consistency, I have made it call c3 and accept all defaults (except `deploy` and `open`). Please advise if this is not the way we want to move forward here.

**Review by IgorMinar — CHANGES_REQUESTED** [review]
I love this clean up, thank you!

These templates have always been something that was of questionable value but we didn't have a product buy-in to remove them.

I do have a concern about the way the test assertion is written and from the PR it's unclear how create-cloudflare is installed when `wrangler init` is invoked - are we ensuring that we end up using the latest create-cloudflare version either via  npx `@latest` or by relying on the auto-update that is built into create-cloudflare.

File: .changeset/gold-toys-smell.md
Comment (IgorMinar) [reply .changeset/gold-toys-smell.md:2]: good catch. thank you! @irvinebroque can you please address this?

File: packages/wrangler/src/__tests__/generate.test.ts
Comment (RamIdeas) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: I think this is because 2.5.0 was the version that included the auto update

Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: I think you can't use semver wildcard here because `toHaveBeenCalledWith` doesn't know what that is, maybe a regular expression instead?

Comment (RamIdeas) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: I'm saying we can make `wrangler generate` run `npm create cloudflare@^2.5.0` via execa. This test would then assert execa was literally called with "cloudflare@^2.5.0". No need for semver knowledge.

Comment (andyjessop) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: We rely on the `getC3CommandFromEnv` function, [which currently defaults to that version](https://github.com/cloudflare/workers-sdk/blob/aa32cb6ae7570c4dd7e1ed6dcc5df3e12bc9ce05/packages/wrangler/src/environment-variables/misc-variables.ts#L29). I'm not sure exactly why we do that, but I used that here for consistency with [how the C3 delegation in `init` is implemented](https://github.com/cloudflare/workers-sdk/blob/aa32cb6ae7570c4dd7e1ed6dcc5df3e12bc9ce05/packages/wrangler/src/init.ts#L202).

File: .changeset/gold-toys-smell.md
Comment (Cherry) [reply .changeset/gold-toys-smell.md:2]: With that, I think the biggest differences will be the "non-interactive mode", but I'd recommend running through all the differences I outlined in https://github.com/cloudflare/workers-sdk/issues/6459 with various use-cases (and perhaps not removing the `generate` tests?) to ensure no breaking changes accidentally land here.

`clones a cloudflare template across drives` was a test I contributed, for example, and ensuring that doesn't regress would still be a good idea.

File: packages/wrangler/src/__tests__/generate.test.ts
Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: I see. thanks @RamIdeas and @andyjessop for the info.

I think https://github.com/cloudflare/workers-sdk/blob/aa32cb6ae7570c4dd7e1ed6dcc5df3e12bc9ce05/packages/wrangler/src/environment-variables/misc-variables.ts#L29 is problematic, because it hardcodes 2.5.0, which means that c3 will download at 2.5.0 and execute just to realize that it's out of date, and use the special auto-update mechanism to update to the latest: https://github.com/cloudflare/workers-sdk/blob/dff8d44f4e47b746b9b1fa276094e1dc4c4f906b/packages/create-cloudflare/src/cli.ts#L41-L65

This double download could be avoided if wrangler invoked create-cloudflare@latest instead of create-cloudflare@2.5.0.

PS: in general I really don't like unpinned dependencies, but with create-cloudflare we really want people to use the latest all the time, so this is an exception to the rule and `@latest` is desirable here.

Comment (Cherry) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: Alternatively, though another network request, you could hit the npm API via `https://registry.npmjs.org/create-cloudflare` or similar, and then semver match against the latest published `2.x` version, if you want to avoid potential breaking changes with people that might use older versions in the future.

Comment (penalosa) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: The reason we didn't initially use latest was to avoid problems if `create-cloudflare` ever had a `3.x` breaking change. We initially tried to use `create-cloudflare@2`, but that didn't work as expected to install the latest `2.x` version. If we're confident C3 will never have a breaking change that will affect this usage maybe `@latest` is fine in this case?

Comment (penalosa) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: I wonder if something like `create cloudflare@"<3.0.0"` would work?

Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: @penalosa you are right! `cloudflare@"<3.0.0"` is the best!

Comment (RamIdeas) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: `@^2.5.0` would resolve the same way (to the latest v2.x.x unless a version above 2.5.0 is cached in which case it would use that and auto update)

In fact, `^2.5.0` is better than `<3.0.0` because it avoids using a cached version older than 2.5.0 which would not autoupdate

Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: hmm.. so thinking about this more, we are likely to encounter caching issue no matter what unless we use @latest which we shouldn't as mentioned earlier because don't want to prevent auto-update to 3.0.0. 

Here is the summary of all the proposal as far as I can tell:

| specifier | status | stale npm cache hit => auto-update | npm cache miss => download latest | double download | protects against uncontrolable create-cloudflare@3.0.0 updates | protects against use of stale create-cloudflare@2.5.0 without auto-update | acceptable solution |
|---------|--------|-------------------------------------|-------------------------------------|-------------------|------------------|------------------|-----------------|
| `@2.5.0` | currently in wrangler | YES ✅  | NO ❌  | YES 😭 | YES ✅  | YES ✅  | NO ❌  |
| `@^2.5.0`  | proposed by @RamIdeas  | YES ✅  | YES ✅  | NO ✅  | YES ✅    | YES ✅    | YES ✅  |
| `@<3.0.0`  | proposed by @penalosa   | YES ✅  | YES ✅  | NO ✅  | YES ✅    | NO ❌  | NO ❌ |
| `@latest`  | proposed by @IgorMinar  | YES ✅  | YES ✅  | NO ✅   | NO ❌ YES ✅    | NO ❌ |  NO ❌ |

If this looks right to y'all then I think @RamIdeas's proposal to go with `@^2.5.0` seems the best. Do we agree? If so can someone please help make this happen?

Comment (RamIdeas) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: PR here https://github.com/cloudflare/workers-sdk/pull/6615

File: .changeset/gold-toys-smell.md
Comment (andyjessop) [reply .changeset/gold-toys-smell.md:2]: The deleted tests are both related to using a named template, which is now covered in [this test](https://github.com/cloudflare/workers-sdk/pull/6415/files#diff-79dc388dc2993d1c5fa4b5c5afc48f26cdf2079685eb491e9aa931af37b4bf0cR58) as it delegates to C3.

File: packages/wrangler/src/__tests__/generate.test.ts
Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: Thank you @RamIdeas and @andyjessop for fixing this!

Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: Is it still expected that the version number remains 2.5.0 in the test and hardcoded? is it because it's mocked out?

Comment (irvinebroque) [reply packages/wrangler/src/__tests__/generate.test.ts:64]: that's how i saw it was before this pr, was also curious how/why it worked this way. nice catch on https://github.com/cloudflare/workers-sdk/pull/6615

Comment (andyjessop) [reply packages/wrangler/src/__tests__/generate.test.ts]: I've updated it now to use `^2.5.0`.

Comment (IgorMinar) [reply packages/wrangler/src/__tests__/generate.test.ts]: great. thanks for the update.

**Review by petebacondarwin — APPROVED** [review]
LGTM

**Review by IgorMinar — APPROVED** [review]
Lgtm for the create-cloudflare versioning issue.

The rest looks good, with one exception that I can't review for on the phone - is this a semver breaking change or did we make the params passing from wrangler to c3 backwards compatible?

I don't want block this PR since I can't review it now, just bringing this up for other reviewers and @irvinebroque to look into. Thank you!

**Comment by Cherry** [comment]
All of my examples at https://github.com/cloudflare/workers-sdk/issues/6459 seem to be using the old code-path for `generate`, so I don't think this is (generally) semver breaking anymore with old `generate` URLs, thanks all. I appreciate the attention given here to ensure this didn't introduce major breaking changes to existing command behaviour in v3.

**Comment by IgorMinar** [comment]
Thanks @andyjessop for the explanation, and @Cherry for additional info and pointer to https://github.com/cloudflare/workers-sdk/issues/6459 - which I believe we can close as soon as this PR is merged, right?

I've reviewed this case and here is my decision in a few bullet points:

- this is a breaking change, and at odds with semver, however...
- we don't really understand the real world impact of this breaking change because we don't have analytics that would tell us how often wrangler generate is invoked with a template name option
- my estimate based on some research that the impact is likely close to 0 (I couldn't find docs or other references where we would instruct users to use `wrangler generate` with a template name anywhere except for a few 3rd party blog posts)
- @irvinebroque has initiated this breaking change and approved it from the product perspective, so we have no blockers there
- it is however wrong and unexpected that if a user calls `wrangler generate my-worker some-template` we give them c3 flow, and ignore the template name — this is a bug that we should address it, however not today
- it's Friday, we should get this change in because overall it's valuable and Monday is feature freeze for bday week
  - we already believe the impact of the breaking change is minimal, we can live with the bug described in the previous point for a few days

Next steps:
- [ ] merge this now as is and release it
- [ ] file a bug describing the issue above
- [ ] fix the bug next week by throwing a descriptive error when the user invokes `wrangler generate` with the template name option, and tell them that this feature is not available and instead they can just use generic c3 flow. Or even prompt them if they'd like us to invoke for them instead.
- [ ] escalate the analytics issue as a blocker for proper decision making which is causing us to spin in circles of endless discussions about what might or might not be the usage of various features. this is a major waste and productivity drag on all of us.


that's all.. does this all matter? in terms of user impact likely not, but if we claim that we want to change our culture and improve DX across the board, then this is the level of attention we need to pay to these changes, and we should implement this level of polish (the bug fix described above) in all the scenarios where we are unsure about the real world impact. Only in cases where we believe that (and have data to prove it) the impact is 0, we can skip the polish.

**Comment by IgorMinar** [comment]
and I see that @penalosa already merged this. thank you!

**Pull request merged**
